### PR TITLE
Delete a user's public key via admin api (closes #3014)

### LINKED ISSUE #3014: Allow admin to delete user keys via API
- Gitea version (or commit ref): 6ad4990
- Git version: 2.7.4
- Operating system: Ubuntu 16.04
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [X] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [X] Not relevant
- Log gist:

## Description
Right now, admins can create keys on behalf of an user, but they can't delete them. (via API)
I suggest that we also support the `DELETE` operation.

From a bigger perspective, is there any reason why the `/user` endpoints where implemented instead of `/users/{username}`? (I guess they were already implemented in gogs)

It's easier for an admin to manage it's users via `/users/{username}`. It can introduce some overhead and confusion if for each resource type endpoint will exist an equivalent admin endpoint (DELETE: `/users/{username}/keys/{id}` -> `/admin/users/{username}/keys/{id}`).

### PR DESCRIPTION
**Pull request by vtemian**
Let admin delete a user's key via API

### CODE REVIEW

**Comment by lafriks** [comment]
Can you also add test that adds and later deletes that key?
Also test for deleting not existing key and test deleting key with user that does not have rights to do that would be great

**Comment by vtemian** [comment]
@lafriks done!

File: routers/api/v1/admin/user.go [hunk 1]
```diff
@@ -236,3 +236,55 @@ func CreatePublicKey(ctx *context.APIContext, form api.CreateKeyOption) {
...
+	// parameters:
+	// - name: username
+	//   in: path
+	//   description: username of user
+	//   type: string
+	//   required: true
+	// - name: id
+	//   in: path
+	//   description: key's id to delete
+	//   type: string
```
Comment (ethantkoenig) [inline routers/api/v1/admin/user.go]: type: integer

File: routers/api/v1/admin/user.go [hunk 2]
```diff
@@ -236,3 +236,55 @@ func CreatePublicKey(ctx *context.APIContext, form api.CreateKeyOption) {
...
+	// - application/json
+	// parameters:
+	// - name: username
+	//   in: path
+	//   description: username of user
+	//   type: string
+	//   required: true
+	// - name: id
+	//   in: path
+	//   description: key's id to delete
```
Comment (ethantkoenig) [inline routers/api/v1/admin/user.go]: nit: "id of the key to delete"

File: routers/api/v1/admin/user.go [hunk 3]
```diff
@@ -236,3 +236,55 @@ func CreatePublicKey(ctx *context.APIContext, form api.CreateKeyOption) {
 	}
 	user.CreateUserPublicKey(ctx, form, u.ID)
 }
+
+// DeleteUserPublicKey api for deleting a user's public key
+func DeleteUserPublicKey(ctx *context.APIContext) {
+	// swagger:operation DELETE /admin/users/{username}/keys/{id} admin adminDeleteUserPublicKey
+	// ---
+	// summary: Delete a user's public key on behalf of a user
```
Comment (ethantkoenig) [inline routers/api/v1/admin/user.go]: nit: "Delete a user's public key". Don't need to repeat "user" twice.

File: routers/api/v1/admin/user.go [hunk 4]
```diff
@@ -236,3 +236,55 @@ func CreatePublicKey(ctx *context.APIContext, form api.CreateKeyOption) {
...
+	//   "404":
+	//     "$ref": "#/responses/notFound"
+	//   "403":
+	//     "$ref": "#/responses/forbidden"
+	u := user.GetUserByParams(ctx)
+	if ctx.Written() {
+		return
+	}
+
+	if _, err := models.GetPublicKeyByID(ctx.ParamsInt64(":id")); err != nil {
```
Comment (ethantkoenig) [inline routers/api/v1/admin/user.go]: It would be nice if we didn't have to call `GetPublicKeyByID` before calling `DeletePublicKey`, since `DeletePublicKey` already calls `GetPublicKeyByID`. Perhaps it's worthwhile to have `DeletePublicKey` return a `ErrKeyNotExist` when you try to delete a non-existent key?

File: integrations/api_admin_test.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	session.MakeRequest(t, req, http.StatusNoContent)
+	models.AssertNotExistsBean(t, &models.PublicKey{ID: newPublicKey.ID})
+}
+
+func TestAPIAdminDeleteMissingSSHKey(t *testing.T) {
+	prepareTestEnv(t)
+	// user1 is an admin user
+	session := loginUser(t, "user1")
+
+	req := NewRequestf(t, "DELETE", "/api/v1/admin/users/user1/keys/99999")
```
Comment (ethantkoenig) [inline integrations/api_admin_test.go]: nit: please use `models.NonexistentID` instead of 99999

**Review by ethantkoenig — CHANGES_REQUESTED** [review]
Mostly looks good, just a few comments

File: routers/api/v1/admin/user.go (hunk 4 above)
Comment (lafriks) [reply routers/api/v1/admin/user.go]: That function is used only in one file, so additional checks for that could be added for that error

**Comment by vtemian** [comment]
@ethantkoenig done

File: public/swagger.v1.json [hunk 6]
```diff
@@ -36,6 +36,7 @@
         ],
         "summary": "Create a user",
         "operationId": "adminCreateUser",
+        "security": null,
```
Comment (lafriks) [inline public/swagger.v1.json:39]: Why these changes?

**Review by lafriks — CHANGES_REQUESTED** [review]

File: public/swagger.v1.json (hunk 6 above)
Comment (vtemian) [reply public/swagger.v1.json:39]: I've run `swagger generate spec` and this was the output. I didn't want to manually change it since it's easier to maintain this way.

**Review by lafriks — APPROVED** [review]

**Review by ethantkoenig — APPROVED** [review]

**Comment by ethantkoenig** [comment]
LGTM

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
LGTM
